src/connectors/infinite-hits/connectInfiniteHits.ts

```diff
@@ -178,107 +176,116 @@ const connectInfiniteHits: InfiniteHitsConnector = function connectInfiniteHits(
+          currentState.facetsRefinements = filterEmptyRefinements(
+            currentState.facetsRefinements
+          );
+          currentState.hierarchicalFacetsRefinements = filterEmptyRefinements(
+            currentState.hierarchicalFacetsRefinements
+          );
+          currentState.disjunctiveFacetsRefinements = filterEmptyRefinements(
+            currentState.disjunctiveFacetsRefinements
+          );
+          currentState.numericRefinements = filterEmptyRefinements(
+            currentState.numericRefinements
+          );
 
-        results.hits = addQueryID(results.hits, results.queryID);
+          if (!isEqual(currentState, prevState)) {
+            cachedHits = cache.read({ state }) || {};
+            prevState = currentState;
+          }
 
-        results.hits = transformItems(results.hits);
+          if (escapeHTML && results.hits.length > 0) {
+            results.hits = escapeHits(results.hits);
+          }
+          const initialEscaped = (results.hits as any).__escaped;
 
-        // Make sure the escaped tag stays after mapping over the hits.
-        // This prevents the hits from being double-escaped if there are multiple
-        // hits widgets mounted on the page.
-        (results.hits as any).__escaped = initialEscaped;
+          results.hits = addAbsolutePosition(
+            results.hits,
+            results.page,
+            results.hitsPerPage
+          );
 
-        if (cachedHits === undefined) {
-          cachedHits = cache.read({ state }) || {};
-        }
+          results.hits = addQueryID(results.hits, results.queryID);
 
-        if (cachedHits![page] === undefined) {
-          cachedHits![page] = results.hits;
-          cache.write({ state, hits: cachedHits });
-        }
+          results.hits = transformItems(results.hits);
 
-        const isFirstPage = getFirstReceivedPage() === 0;
-        const isLastPage = results.nbPages <= results.page + 1;
+          // Make sure the escaped tag stays after mapping over the hits.
+          // This prevents the hits from being double-escaped if there are multiple
+          // hits widgets mounted on the page.
+          (results.hits as any).__escaped = initialEscaped;
 
-        renderFn(
-          {
-            hits: extractHitsFromCachedHits(cachedHits!),
-            results,
-            showPrevious,
-            showMore,
-            isFirstPage,
-            isLastPage,
-            instantSearchInstance,
-            widgetParams,
-          },
-          false
-        );
+          if (cachedHits === undefined) {
+            cachedHits = cache.read({ state }) || {};
+          }
+
+          if (cachedHits![page] === undefined) {
+            cachedHits![page] = results.hits;
+            cache.write({ state, hits: cachedHits });
+          }
+
+          hits = extractHitsFromCachedHits(cachedHits!);
+        }
+
+        return {
+          hits,
+          results,
+          showPrevious: getShowPrevious(helper),
```

Comment:
these references are no longer stable now I think

Reply:
what do you mean by no longer stable? could you explain more?

Reply:
if you do addEventListener on an early render, the function reference isn't the same as another one, meaning that removeEventListener wouldn't work I think (although I always get confused about this)

Reply:
Understood, that's very error prone. I updated the tests to check that we get the same function at the first and second renders.